pkg/cstor/pv_operation.go

```diff
@@ -58,7 +59,53 @@ func (p *Plugin) updateVolCASInfo(data []byte, volumeID string) error {
 	return nil
 }
 
+// restoreVolumeFromCloud restore remote snapshot for the given volume
+// Note: cstor snapshots are incremental in nature, so restore will be executed
+// from base snapshot to incremental snapshot 'vol.backupName' if p.restoreAllSnapshots is set
+// else restore will be performed for the given backup only.
 func (p *Plugin) restoreVolumeFromCloud(vol *Volume) error {
+	var (
+		snapshotList []string
+		err          error
+	)
+
+	targetBackupName := vol.backupName
+
+	if p.restoreAllSnapshots {
+		// We are restoring from base backup to targeted Backup
+		snapshotList, err = p.cl.GetSnapListFromCloud(vol.snapshotTag, p.getScheduleName(vol.backupName))
+		if err != nil {
+			return err
+		}
+	} else {
+		// We are restoring only given backup
+		snapshotList = []string{targetBackupName}
+	}
+
+	// snapshots are created using timestamp, we need to sort it in ascending order
+	sort.Strings(snapshotList)
+
+	for _, snap := range snapshotList {
+		p.Log.Infof("Restoring snapshot=%s", snap)
+
+		vol.backupName = snap
```

Comment:
this is overriding 'vol' which is input to this method. That might be not intended behavior.
can we make a copy of 'vol' before changing it?

Reply:
`backupName` act as argument for restore. Caller uses `vol` pointer for update on status. I'll change this function to take backupName as arg and update vol.backupName for restore.